### PR TITLE
Add -Wextra flag

### PR DESCRIPTION
**Pull request by masterleinad**
Addresses #2796. This works for me locally.
Some things still to address:
- add `-Wextra` to the documentation of flags used (and `Makefile`, `testing_scripts`)
- ~~deal with the `fallthrough` warning for `containers/src/impl/Kokkos_Functional_impl.hpp:109` (hopefully removing `-Wno-implicit-fallthrough`.~~
- ~~find a better workaround in `core/src/Kokkos_View.hpp`~~
- make sure ignoring all unused arguments is intended indeed

### CODE REVIEW

File: containers/src/Kokkos_DynRankView.hpp [hunk 1]
```diff
@@ -2143,7 +2143,7 @@ inline void resize(DynRankView<T, P...>& v,
   static_assert(Kokkos::ViewTraits<T, P...>::is_managed,
                 "Can only resize managed views");
 
-  drview_type v_resized(v.label(), n0, n1, n2, n3, n4, n5, n6);
+  drview_type v_resized(v.label(), n0, n1, n2, n3, n4, n5, n6, n7);
```
Comment (calewis) [inline containers/src/Kokkos_DynRankView.hpp]: One issue fixed due to this warning :)

File: containers/src/Kokkos_DynRankView.hpp [hunk 2]
```diff
@@ -2170,7 +2170,7 @@ inline void realloc(DynRankView<T, P...>& v,
   const std::string label = v.label();
 
   v = drview_type();  // Deallocate first, if the only view to allocation
-  v = drview_type(label, n0, n1, n2, n3, n4, n5, n6);
+  v = drview_type(label, n0, n1, n2, n3, n4, n5, n6, n7);
```
Comment (crtrott) [inline containers/src/Kokkos_DynRankView.hpp]: We need a unit test which catches this mistake.

File: core/perf_test/PerfTest_CustomReduction.cpp [hunk 3]
```diff
@@ -104,7 +104,7 @@ void custom_reduction_test(int N, int R, int num_trials) {
                   Scalar t_max = Scalar(0);
                   Kokkos::parallel_reduce(
                       Kokkos::ThreadVectorRange(team, 32),
-                      [&](const int& k, Scalar& max_) {
+                      [&](const int& k, Scalar& /*max_*/) {
```
Comment (crtrott) [inline core/perf_test/PerfTest_CustomReduction.cpp]: uhoh no. This is a bug in the test. the argument should be used!!

File: core/src/Cuda/Kokkos_CudaSpace.cpp [hunk 4]
```diff
@@ -789,6 +789,8 @@ SharedAllocationRecord<Kokkos::CudaHostPinnedSpace, void>
 // Iterate records to print orphaned memory ...
 void SharedAllocationRecord<Kokkos::CudaSpace, void>::print_records(
     std::ostream &s, const Kokkos::CudaSpace &, bool detail) {
+  (void)s;
+  (void)detail;
```
Comment (crtrott) [inline core/src/Cuda/Kokkos_CudaSpace.cpp]: from experience we know that this doesn't work on all compilers. but we can try.

File: core/src/Kokkos_View.hpp [hunk 5]
```diff
@@ -830,10 +830,10 @@ class View : public ViewTraits<DataType, Properties...> {
...
 #else
 
-#define KOKKOS_IMPL_VIEW_OPERATOR_VERIFY(ARG) \
-  View::template verify_space<                \
-      Kokkos::Impl::ActiveExecutionMemorySpace>::check();
-
+#define KOKKOS_IMPL_VIEW_OPERATOR_VERIFY(ARG)             \
+  View::template verify_space<                            \
+      Kokkos::Impl::ActiveExecutionMemorySpace>::check(); \
+  [](...) {} ARG;
```
Comment (crtrott) [inline core/src/Kokkos_View.hpp]: What is this??

File: core/src/impl/Kokkos_ViewMapping.hpp [hunk 6]
```diff
@@ -2113,7 +2113,8 @@ struct ViewStride;
 
 template <>
 struct ViewStride<0> {
-  enum { S0 = 0, S1 = 0, S2 = 0, S3 = 0, S4 = 0, S5 = 0, S6 = 0, S7 = 0 };
+  static constexpr size_t S0 = 0, S1 = 0, S2 = 0, S3 = 0, S4 = 0, S5 = 0,
+                          S6 = 0, S7 = 0;
```
Comment (crtrott) [inline core/src/impl/Kokkos_ViewMapping.hpp]: Uhoh. No please not for now. We had compilers where static constexpr still doesn't work when accessed on the GPU in particular with OpenMP Target

**Comment by crtrott** [comment]
Generally I think Wextra gives more than we may be able to handle. There is really funky stuff in that.

**Comment by crtrott** [comment]
This one from the CUDA 9.2 NVCC build is nice:
```
cc1plus: error: unrecognized command line option '-Wno-implicit-fallthrough' [-Werror]
```

File: core/src/Cuda/Kokkos_CudaSpace.cpp (hunk 4 above)
Comment (dalg24) [reply core/src/Cuda/Kokkos_CudaSpace.cpp]: How about `std::ignore`?

Comment (masterleinad) [reply core/src/Cuda/Kokkos_CudaSpace.cpp]: > from experience we know that this doesn't work on all compilers. but we can try.

Do you have any example we should look into?

> How about `std::ignore`?

That won't work in device code so we can't use it everywhere as a replacement.

File: core/src/Kokkos_View.hpp (hunk 5 above)
Comment (masterleinad) [reply core/src/Kokkos_View.hpp]: That was the quickest workaround I could think of to ignore arbitrary arguments. I changed the workaround now.

**Comment by masterleinad** [comment]
AFAICT (and apart from the question of `(void)` working for all relevant compilers), only the `[[fallthrough]]` handling is still a bit challenging.

File: core/src/impl/Kokkos_ViewMapping.hpp (hunk 6 above)
Comment (masterleinad) [reply core/src/impl/Kokkos_ViewMapping.hpp]: I found another workaround.

File: core/src/Cuda/Kokkos_CudaSpace.cpp (hunk 4 above)
Comment (dalg24) [reply core/src/Cuda/Kokkos_CudaSpace.cpp]: This function right here is not marked `__host__ __device__`

Comment (masterleinad) [reply core/src/Cuda/Kokkos_CudaSpace.cpp]: I know. If possible, I would go with a common style. That's why I wrote "everywhere".

**Comment by masterleinad** [comment]
Looks like this is finally passing.

**Comment by calewis** [comment]
> Generally I think Wextra gives more than we may be able to handle. There is really funky stuff in that.

Agreed, also Wextra isn't stable if we add new compliers.  But I do think we should include unused parameters and unused variables, since we caught at least 2 bugs with the parameters one.

File: containers/src/impl/Kokkos_Functional_impl.hpp [hunk 7]
```diff
@@ -105,16 +105,18 @@ uint32_t MurmurHash3_x86_32(const void* key, int len, uint32_t seed) {
...
-    case 3: k1 ^= tail[2] << 16;
-    case 2: k1 ^= tail[1] << 8;
-    case 1:
-      k1 ^= tail[0];
-      k1 *= c1;
-      k1 = rotl32(k1, 15);
-      k1 *= c2;
-      h1 ^= k1;
-  };
+  const int len_and_3 = len & 3;
```
Comment (crtrott) [inline containers/src/impl/Kokkos_Functional_impl.hpp]: uhm is this the same thing? for example `k1^=tail[1] << 8` is now executed when `len&3==3` which wasn't true before...

File: containers/unit_tests/TestOffsetView.hpp [hunk 8]
```diff
@@ -127,7 +127,7 @@ void test_offsetview_construction(unsigned int size) {
   typedef typename range_type::point_type point_type;
 
   range_type rangePolicy2D(point_type{{ovmin0, ovmin1}},
-                           point_type{{ovend0, ovend1}});
+                           point_type{{ovend0, ovend1}}, point_type{{0, 0}});
```
Comment (crtrott) [inline containers/unit_tests/TestOffsetView.hpp]: what is this point_type{{0,0}} here added for? Also doesn't this kind of explicitly assume that `{0,0}` means default for tiling?

File: containers/unit_tests/TestOffsetView.hpp [hunk 9]
```diff
@@ -195,7 +195,8 @@ void test_offsetview_construction(unsigned int size) {
     typedef typename range3_type::point_type point3_type;
 
     range3_type rangePolicy3DZero(point3_type{{0, 0, 0}},
-                                  point3_type{{extent0, extent1, extent2}});
+                                  point3_type{{extent0, extent1, extent2}},
+                                  point3_type{{0, 0, 0}});
```
Comment (crtrott) [inline containers/unit_tests/TestOffsetView.hpp]: Same as above.

File: containers/unit_tests/TestOffsetView.hpp [hunk 10]
```diff
@@ -209,7 +210,8 @@ void test_offsetview_construction(unsigned int size) {
     range3_type rangePolicy3D(
         point3_type{{begins[0], begins[1], begins[2]}},
         point3_type{
-            {begins[0] + extent0, begins[1] + extent1, begins[2] + extent2}});
+            {begins[0] + extent0, begins[1] + extent1, begins[2] + extent2}},
+        point3_type{{0, 0, 0}});
```
Comment (crtrott) [inline containers/unit_tests/TestOffsetView.hpp]: same as above.

File: containers/unit_tests/TestOffsetView.hpp [hunk 11]
```diff
@@ -457,7 +459,8 @@ void test_offsetview_subview(unsigned int size) {
       const int e0 = offsetSubview.end(0);
       const int e1 = offsetSubview.end(1);
 
-      range_type rangeP2D(point_type{{b0, b1}}, point_type{{e0, e1}});
+      range_type rangeP2D(point_type{{b0, b1}}, point_type{{e0, e1}},
+                          point_type{{0, 0}});
```
Comment (crtrott) [inline containers/unit_tests/TestOffsetView.hpp]: same as above.

File: core/src/Kokkos_View.hpp [hunk 12]
```diff
@@ -821,21 +821,11 @@ class View : public ViewTraits<DataType, Properties...> {
...
   Kokkos::Impl::view_verify_operator_bounds<typename traits::memory_space> ARG;
 
-#else
-
-#define KOKKOS_IMPL_VIEW_OPERATOR_VERIFY(ARG) \
-  View::template verify_space<                \
-      Kokkos::Impl::ActiveExecutionMemorySpace>::check();
-
-#endif
-
```
Comment (crtrott) [inline core/src/Kokkos_View.hpp]: Why this change?

File: core/src/impl/Kokkos_Atomic_Generic.hpp [hunk 13]
```diff
@@ -178,7 +178,7 @@ KOKKOS_INLINE_FUNCTION T atomic_fetch_oper(
 
 template <class Oper, typename T>
 KOKKOS_INLINE_FUNCTION T atomic_oper_fetch(
-    const Oper& op, volatile T* const dest,
+    const Oper& /*op*/, volatile T* const dest,
```
Comment (crtrott) [inline core/src/impl/Kokkos_Atomic_Generic.hpp]: the mistake is actually that it didn't call op.apply down there.

File: core/src/impl/Kokkos_ViewMapping.hpp [hunk 14]
```diff
@@ -2043,19 +2043,23 @@ struct ViewOffset<
...
+      right_stride_is_1 = (rhs.m_stride.S3 == 1);
+    else if (dimension_type::rank == 5)
+      right_stride_is_1 = (rhs.m_stride.S4 == 1);
+    else if (dimension_type::rank == 6)
+      right_stride_is_1 = (rhs.m_stride.S5 == 1);
+    else if (dimension_type::rank == 7)
+      right_stride_is_1 = (rhs.m_stride.S6 == 1);
+    else if (dimension_type::rank == 8)
+      right_stride_is_1 = (rhs.m_stride.S7 == 1);
+
```
Comment (crtrott) [inline core/src/impl/Kokkos_ViewMapping.hpp]: There is a reason that we wrote this as a big ternary operator. Can we verify that this produces the same code instructions at the end?

File: core/unit_test/TestMDRange.hpp [hunk 15]
```diff
@@ -644,7 +644,8 @@ struct TestMDRange_2D {
               range_type;
       typedef typename range_type::point_type point_type;
 
-      range_type range(point_type{{0, 0}}, point_type{{N0, N1}});
+      range_type range(point_type{{0, 0}}, point_type{{N0, N1}},
+                       point_type{{0, 0}});
```
Comment (crtrott) [inline core/unit_test/TestMDRange.hpp]: as above I am not happy about explicitly adding the tiling argument here.

File: core/unit_test/TestMDRange.hpp [hunk 16]
```diff
@@ -1232,7 +1232,8 @@ struct TestMDRange_3D {
           range_type;
       typedef typename range_type::point_type point_type;
 
-      range_type range(point_type{{0, 0, 0}}, point_type{{N0, N1, N2}});
+      range_type range(point_type{{0, 0, 0}}, point_type{{N0, N1, N2}},
+                       point_type{{0, 0, 0}});
```
Comment (crtrott) [inline core/unit_test/TestMDRange.hpp]: as above.

File: core/unit_test/TestMDRange.hpp [hunk 17]
```diff
@@ -1868,7 +1869,8 @@ struct TestMDRange_4D {
           range_type;
       typedef typename range_type::point_type point_type;
 
-      range_type range(point_type{{0, 0, 0, 0}}, point_type{{N0, N1, N2, N3}});
+      range_type range(point_type{{0, 0, 0, 0}}, point_type{{N0, N1, N2, N3}},
+                       point_type{{0, 0, 0, 0}});
```
Comment (crtrott) [inline core/unit_test/TestMDRange.hpp]: as above.

File: core/unit_test/TestMDRange.hpp [hunk 18]
```diff
@@ -2437,7 +2440,8 @@ struct TestMDRange_5D {
       typedef typename range_type::point_type point_type;
 
       range_type range(point_type{{0, 0, 0, 0, 0}},
-                       point_type{{N0, N1, N2, N3, N4}});
+                       point_type{{N0, N1, N2, N3, N4}},
+                       point_type{{0, 0, 0, 0, 0}});
```
Comment (crtrott) [inline core/unit_test/TestMDRange.hpp]: as above

File: core/unit_test/TestMDRange.hpp [hunk 19]
```diff
@@ -3030,7 +3034,8 @@ struct TestMDRange_6D {
       typedef typename range_type::point_type point_type;
 
       range_type range(point_type{{0, 0, 0, 0, 0, 0}},
-                       point_type{{N0, N1, N2, N3, N4, N5}});
+                       point_type{{N0, N1, N2, N3, N4, N5}},
+                       point_type{{0, 0, 0, 0, 0, 0}});
```
Comment (crtrott) [inline core/unit_test/TestMDRange.hpp]: as above.

File: containers/unit_tests/TestOffsetView.hpp (hunk 8 above)
Comment (masterleinad) [reply containers/unit_tests/TestOffsetView.hpp]: Yes, this is for tiling and the tiling argument is zero-initialized by default. Only `gcc-4.8.4` complains here with `-Wmissing-field-initializers`. If we don't want all of `-Wextra`, I would just drop this change.

Comment (crtrott) [reply containers/unit_tests/TestOffsetView.hpp]: Yeah we don't do this, in particular not for gcc 4.8.4

File: containers/src/Kokkos_DynRankView.hpp (hunk 2 above)
Comment (masterleinad) [reply containers/src/Kokkos_DynRankView.hpp]: Allowing `DynRankView` to use 8 dimensions seems like a bigger change and should be in a different pull request if we decide we want it.
I am happy to just fix this issue here, though, and add a small test.

Comment (ndellingwood) [reply containers/src/Kokkos_DynRankView.hpp]: I think DynRankView reserves the 8th rank for Sacado

Comment (masterleinad) [reply containers/src/Kokkos_DynRankView.hpp]: Are you saying the correct approach would be to restrict this constructor to seven dimensions as well?

File: containers/src/impl/Kokkos_Functional_impl.hpp (hunk 7 above)
Comment (masterleinad) [reply containers/src/impl/Kokkos_Functional_impl.hpp]: AFAICT, it was supposed to fall through.

File: core/src/Kokkos_View.hpp (hunk 12 above)
Comment (masterleinad) [reply core/src/Kokkos_View.hpp]: I couldn't avoid the warning about the passed arguments being unused differently.

**Comment by masterleinad** [comment]
Summary:
- Most of the places fixed here are related to `-Wunused-parameter`.
- Enabling `-Wunused-parameter` made us aware of some real issues and seems to be supported by all the compilers we check for.
- The warnings enabled by `-Wextra` vary quite a bit between compilers. One example would be `-Wimplicit-fallthrough` which is rather annoying to fix using a solution that works for all the compilers we test for.
- `-Wextra` enables some warnings that trigger compiler bugs with `gcc-4.8.4`.

Can we conclude that we don't want all of `-Wextra` but that it makes sense to add `-Wunused-parameter`?

**Comment by masterleinad** [comment]
We decided that `-Wextra` is going too far for us to compile with all supported compilers warning-free.
